### ISSUE TITLE
Collaboration

### ISSUE THREAD

**Issue by disordinary** [bustlelabs/mobiledoc-kit#429]
I was wondering if any thought had been given towards collaboration? 

This is an issue for what I want to use Mobile Doc for, and if there is the appetite in the project to add a collaboration aspect to it, then something that I can tackle. 

Please forgive me if this has already been brought up, I had a look and couldn't find anything.

**Comment by disordinary** [comment]
A lot of the required work can be done simply by using the hooks available. However, it would appear that history handling in mobile docs will get in the way and require a more invasive change. 

That being the case, other invasive changes should be looked at, weighed, and measured. CRDT is in someways an easier way to implement collaborative text editing than OT based solutions, but because it's a data type it will require modification of the core text handling within MD. And while we're rewriting the way that mobile doc deals with strings internally we might as well add the ability to have multiple cursors on the UI side. 

It's a big job.

**Comment by mixonic** [comment]
@disordinary right, real-time collaboration is definitely a large challenge. We've not put any time toward that goal yet, instead preferring to focus on other goals (rich content embedding, programmatic API etc). Collaborative editing would obviously be a great win.

There are a couple approaches and design decisions to think about. CRDT would be great, but do we make the Mobiledoc format itself a CRDT? What are the implications for rendering speed outside of the editor? I believe some CRDTs grow unbounded in memory over time as well, and I'm uncertain of the implications there.

Currently, the Mobiledoc format explicitly does not contain any data that is editing-specific. Only presentation-specific data is stored. Adding cursor position, and even possible document history, to the persistence format itself would be a significant change. This isn't a blocker, however it does highlight that we may be best served by a two-channel system, an edits format and a document format. Or perhaps not.

@disordinary if you have the bandwidth to work on this and a use-case calling for it, @bantic and I would certainly get on a hangout to talk about approaches to the problem. We can chat on gitter!

**Comment by disordinary** [comment]
Hi @mixonic,

I definitely have the capacity to research the problem and put together a prototype solution.

The thing that I need to get my head around with a CRDT is how it will work with the immutable data structure of mobile doc. One of the issues is that a CRDT cannot delete characters, therefore instead of removing them with an undo they need to be tombstoned, or marked as invisible. And you're right a CRDT will grow without bounds. I think the best implementation would be as a binary-tree, the various editors collaborating on the document can then form a consensus when a section of the CRDT no longer has a cursor placed in it and collapse that branch back into a standard string. 

In someways an OT approach is the cleanest and easiest, at least initially. The OT operations can be stored separately from the history and any external edits that have been lost when an undo occurs replayed. It has the least amount of invasiveness, but is notoriously hard to get right for anything other than the simplest operations, and long term may prove to be the more challenging solution.

Another option is simply to lock a card or section when a user is working on it, and then unlock it when they have finished.

I've sent you an IM on gitter, happy to have a more in-depth conversation.

**Comment by disordinary** [comment]
Hey @mixonic sorry things have gone quiet on this, when we had a chat about collaboration you mentioned that someone else was interested in building a CRDT into the mobiledoc format itself, I'd appreciate it if you could throw their contact details my way and we can talk about possibly sharing the load of implementing that feature. To be honest the Mobiledoc format is about the worst format possible to build a CRDT on top of so I'd love to discuss some of their ideas.

**Comment by mixonic** [comment]
@disordinary I would really like to see us experiment with this, and I'd like to find some support from user organizations to push it forward this winter. I'll ping you on chat or email and we can schedule a call for later this week. 201 Created is working on OSS this week, though we're focusing on an Ember project for the most part.

**Comment by tim-evans** [comment]
I'm gonna stick my head in here and say that our team may be super interested in this. We'd love to grab drink / coffee / lunch and chat to see if this aligns with our team's efforts

**Comment by disordinary** [comment]
Let me know what the outcome of any chat is because I'd be keen to contribute.

From discussions that I've had with @mixonic  and @bantic building a CRDT into the mobiledoc format seems to be what is favored, the challenge being that the mobiledoc format is not particularly conducive to this.

Because this has never progressed past a casual conversation I'd imagine finding alignment would be very easy. What were your requirements and thoughts @tim-evans?

**Comment by rlivsey** [comment]
I'm also very interested in this and keen to help out where possible

**Comment by tim-evans** [comment]
We need to support live blogging for events. That's one of the asks we have

**Comment by disordinary** [comment]
It appears as though no one is going to take a lead on this, this is understandable because the feature is going to be incredibly disruptive to the project, from the format all the way up. 

I have some experience building collaborative editors, but nothing more than toy projects or tools for internal usage. I'm happy to keep nudging this along until someone more qualified arrives.

My original idea was to add the hooks required for collaborative editing into the editor but leave the project otherwise unscathed. The actual collaboration code would be external to Mobiledoc-kit core and would most likely be a lite implementation of an OT like system. 

The main sticking point would be the way that mobiledoc-kit handles history as an array of immutable states. It would require a complete rewrite to be based on the commands that are sent as part of the OT system which is how it's historically been done, or by keeping a list of changes between states which is how prosemirror does it (IIRC). 

The preferred approach would appear to be a CRDT based system, which has numerous advantageous to OT, but is also more invasive to mobiledoc-kit as it would require a complete update of the mobiledoc format - I have some ideas around this, and will do a write-up at some point.
The difference between the two IMHO is that:

- A CRDT system has more upfront work but once it's up and running it should just work without worrying about edge cases. We also get things like real-time collaboration and also conflict-free merging when mobiledoc suffers a split brain (for instance if someone is writing while offline). 

- An OT based system has a conceptually simpler model and should be quicker to get up and running, but is notoriously hard to make reliable. My opinion is if we go down this route it would be for a system that is realtime only with no offline merging.

**Comment by mixonic** [comment]
@disordinary I've love to sync up early next week and chat over ideas for how to structure the effort. I'm still leaning toward CRDT myself. The OT approach feels like it will leave us in a dead end.

We should talk about the design of an API for altering Mobiledoc, and how we could build that library as a stand-alone lib for the Mobiledoc-Kit editor to consume. I hope this would let us decouple the issue of how to decouple Mobiledoc-Kit's UI and AST code from the issue of how to create a CRDT Mobiledoc.

**Comment by karellm** [comment]
I was wondering if there was any update on the topic. I'm currently evaluating mobiledoc-kit for a project and it being a CRDT system would be a huge plus. Thanks!

**Comment by disordinary** [comment]
Hey Karellm, no real update other than we've validated the approach works. It's a huge job and there hasn't really been the need to justify development.

**Comment by disordinary** [comment]
If you join the mobiledoc-kit slack there is a CRDT slack channel where the development talk is happening. It's pretty quiet though, as I said it's a pointless feature to add unless there is a genuine need for it and probably someone who is willing to pay for the engineering time.

**Comment by karellm** [comment]
Thanks I will check that out!
